hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/util/Shell.java

```diff
@@ -89,6 +89,22 @@ public static boolean isJava7OrAbove() {
     return true;
   }
 
+  private static final boolean IS_JAVA9_OR_ABOVE;
+  static {
+    int major = 0;
+    try {
+      // 1.8.0 -> 1, 10-ea+30 -> 10, 18.03 -> 18, 11 -> 11 etc.
+      String version = System.getProperty("java.version");
+      major = Integer.parseInt(version.split("\\.")[0].split("-")[0]);
+    } catch (NumberFormatException ignored) {
+    }
+    IS_JAVA9_OR_ABOVE = major >= 9;
+  }
+
+  public static boolean isJava9OrAbove() {
+    return IS_JAVA9_OR_ABOVE;
+  }
```

Comment:
Given we've already got isJava7OrAbove() tagged as deprecated, we should stop repeating the same mistake.

How about some: `boolean IsJavaVersionAtLeast(int)` which can do checks for versions, store this parsed major in a private static (set the default value to 8, obviously) field, then have the new predicate return true if the param is >= the cached value

Reply:
Yes, I had exactly the same in mind. Just haven't realized we may now assume 8 as lowest version. Will it still be true if someone decides to backport the fix to previous Hadoop versions?

Reply:
Hadoop trunk is java 8+ only.

* nobody will backport your patch to branch-2; its using Java 8 classes.
* someone could backport an `isJavaVersionAtLeast()` method to branch-2, in which case, they get to extend it.